sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarToGpu.scala

```diff
@@ -107,9 +107,15 @@ object HostColumnarToGpu extends Logging {
     referenceManagers.result().asJava
   }
 
-  def columnarCopy(cv: ColumnVector,
-      b: ai.rapids.cudf.HostColumnVector.ColumnBuilder, rows: Int): Unit = {
-    cv.dataType() match {
+  // Data type is passed explicitly to allow overriding the reported type from the column vector.
+  // There are cases where the type reported by the column vector does not match the data.
+  // See https://github.com/apache/iceberg/issues/6116.
+  def columnarCopy(
+      cv: ColumnVector,
+      b: ai.rapids.cudf.HostColumnVector.ColumnBuilder,
+      dataType: DataType,
+      rows: Int): Unit = {
+    dataType match {
```

Comment:
Do we want a warning or something if `dataType` and `cv.dataType()` do not match? I am not sure about that.

Reply:
I thought about it, but then the plugin would spit out a warning for some Iceberg use-cases, possibly quite often.  That could be scary for users, who are in no position to do anything about it.  Given we know this can happen in a "normal" use-case, I thought it best to leave it out.  Note that we still have some protections if the data doesn't even match the Spark schema since the column will likely throw some type of exception (e.g.: class cast exception) when it is asked to marshal into the wrong type.

Reply:
Filed #7012 to track removing the workaround when no longer necessary.